File: algorithm/python/prob1001-1500/prob1484_medium_Clone-Binary-Tree-With-Random-Pointer.py
class Node:
    def __init__(self, val=0, left=None, right=None, random=None):
        self.val = val
        self.left = left
        self.right = right
        self.random = random

class NodeCopy:
    def __init__(self, val=0, left=None, right=None, random=None):
        self.val = val
        self.left = left
        self.right = right
        self.random = random

class Solution:
    def copyRandomBinaryTree(self, root: 'Node') -> 'NodeCopy':
        if root is None:
            return None
        self.mapping = dict() # 原节点 -> 对应的已拷贝的节点
        return self.dfs(root)

    def dfs(self, node: "Node") -> "NodeCopy":
        new_node = NodeCopy(node.val)
        self.mapping[node] = new_node
        if node.left is not None:
            if node.left in self.mapping:
                new_node.left = self.mapping[node.left]
            else:
                new_node.left = self.dfs(node.left)
        if node.right is not None:
            if node.right in self.mapping:
                new_node.right = self.mapping[node.right]
            else:
                new_node.right = self.dfs(node.right)
        if node.random is not None:
            if node.random in self.mapping:
                new_node.random = self.mapping[node.random]
            else:
                new_node.random = self.dfs(node.random)
        return new_node

File: algorithm/cpp/prob1001-1500/prob1484_medium_Clone-Binary-Tree-With-Random-Pointer.cpp
// prob1484: Clone Binary Tree With Random Pointer

/*
 * https://leetcode-cn.com/problems/clone-binary-tree-with-random-pointer/
 */

/**
 * Definition for a binary tree node.
 */

#include <unordered_map>

using namespace std;

struct Node {
    int val;
    Node *left;
    Node *right;
    Node *random;
    Node() : val(0), left(nullptr), right(nullptr), random(nullptr) {}
    Node(int x) : val(x), left(nullptr), right(nullptr), random(nullptr) {}
    Node(int x, Node *left, Node *right, Node *random) : val(x), left(left), right(right), random(random) {}
};

struct NodeCopy {
    int val;
    NodeCopy *left;
    NodeCopy *right;
    NodeCopy *random;
    NodeCopy() : val(0), left(nullptr), right(nullptr), random(nullptr) {}
    NodeCopy(int x) : val(x), left(nullptr), right(nullptr), random(nullptr) {}
    NodeCopy(int x, NodeCopy *left, NodeCopy *right, NodeCopy *random) : val(x), left(left), right(right), random(random) {}
};


class Solution {
public:
    NodeCopy* copyRandomBinaryTree(Node* root) {
        if(!root) return nullptr;
        return dfs(root);
    }

private:
    unordered_map<Node*, NodeCopy*> mapping;

    NodeCopy* dfs(Node* node)
    {
        NodeCopy *new_node = new NodeCopy(node -> val);
        mapping[node] = new_node;
        if(node -> left)
        {
            if(mapping.count(node -> left) > 0)
                new_node -> left = mapping[node -> left];
            else
                new_node -> left = dfs(node -> left);
        }
        if(node -> right)
        {
            if(mapping.count(node -> right) > 0)
                new_node -> right = mapping[node -> right];
            else
                new_node -> right = dfs(node -> right);
        }
        if(node -> random)
        {
            if(mapping.count(node -> random) > 0)
                new_node -> random = mapping[node -> random];
            else
                new_node -> random = dfs(node -> random);
        }
        return new_node;
    }
};
